build_dependency_tree.py

```diff
@@ -3,76 +3,202 @@
     impls = [file for file in files_in_dir if file.suffix in ('c', 'cpp')]
     return headers, impls
 
 
-def create_single_file_dependency_list(file: Path):
+# NOTE: Do we want to make re.Pattern object a function parameter?
+# KI: No, but also don't want it to be a global variable. This screams classes
+#
+# match either the pattern `<...>` or `"..."`
+include_statement_pattern = re.compile(r"(<.+?>)|(\".+?\")")
+
+
+def create_single_file_dependency_list(file: Union[Path, str]):
     """
-    TODO: Go through header files and implementation files and scan for `#include`
-          keyword, then maybe some regex to extract the file name.
-          For each file we go through, we create a list of the files required by
-          it.
-    NOTE: With C++20 and C++23 we will have to include import statements as well
-          But we leave this for future work
-    
-    :param file: path to C++ file in current project
-    return: returns: a list of the files that `file` depends on
+    Go through header files and implementation files and scan for `#include`
+    keyword, then maybe some regex to extract the file name.
+    For each file we go through, we create a list of the files required by
+    it.
+
+    With C++20 and C++23 we will have to include import statements as well
+    But we leave this for future work
+
+    :param file:        path to C++ file in current project
+    return: returns:    a list of the files that `file` depends on
     """
-    with open(file) as f:
-        # some files may have macros that check for operatoring system or 
+    # For some reason does not drop the <> or " when matching
+    # so we write a helper
+    def extract_header_file_name(text: Optional[str]):
+        if text:
+            return text[1:len(text)-1]
+        else:
+            return None
+
+    with open(str(file.absolute())) as f:
+        # some files may have macros that check for operatoring system or
         # compiler compatibilities. For now we ignore these as well.
-        # We also need to ignore include statements made in comments 
+        # We also need to ignore include statements made in comments
         include_statements = []
-        b_is_block_comment = False # Lines that start with /*
+        b_is_block_comment = False  # Lines that start with /*
         for line in f.readlines():
+            print(f"{line.strip()}")
             if b_is_block_comment:
-                if line.split()[0] == '*/':
+                if '*/' in line:
                     b_is_block_comment = False
                 else:
                     continue
             else:
-                if line.contains('/*'):
+                if '/*' in line:
                     b_is_block_comment = True
                     continue
-                elif line.split() == '//':
+                elif line.replace(' ', '')[0:2] == '//':
+                    continue
+                elif not line:  # empty line
                     continue
```

Comment:
I advise to establish branching logic such that this simple continue conditions are executed implicitly (not satisfying any conditions in the if clauses)

Reply:
When we know that a specific branch with be executed more frequently then another, we want to put it first to avoid unnecessary logic checks. 
Having a is designed that it flows logically to a human, may not make it the most efficient code.
But I have tested that this runs pretty damn quickly even on very large code bases, so I am not too concerned with performance